HeterogeneousCore/SonicTriton/src/TritonClient.cc

```diff
@@ -0,0 +1,405 @@
+
+  //should be true
+  return status;
+}
+
+template <typename Client>
+bool TritonClient<Client>::getResults(const nic::InferContext::Result& result) {
+  bool status = true;
+
+  auto t1 = std::chrono::high_resolution_clock::now();
+  this->output_.resize(nOutput_ * batchSize_, 0.f);
+  for (unsigned i0 = 0; i0 < batchSize_; i0++) {
+    const uint8_t* r0;
+    size_t content_byte_size;
+    status = wrap(result.GetRaw(i0, &r0, &content_byte_size),
+                  "getResults(): unable to get raw for entry " + std::to_string(i0));
+    if (!status)
+      return status;
+    std::memcpy(&this->output_[i0 * nOutput_], r0, nOutput_ * sizeof(float));
+  }
+  auto t2 = std::chrono::high_resolution_clock::now();
+  if (!this->debugName_.empty())
+    edm::LogInfo(this->fullDebugName_) << "Output array time: "
+                                       << std::chrono::duration_cast<std::chrono::microseconds>(t2 - t1).count();
+
+  //should be true
+  return status;
+}
+
+//default case for sync and pseudo async
+template <typename Client>
+void TritonClient<Client>::evaluate() {
+  //in case there is nothing to process
+  if (batchSize_ == 0) {
+    this->finish(true);
+    return;
+  }
+
+  //common operations first
+  bool status = setup();
+  if (!status) {
+    this->finish(false);
+    return;
+  }
+
+  // Get the status of the server prior to the request being made.
+  const auto& start_status = getServerSideStatus();
+
+  //blocking call
+  auto t1 = std::chrono::high_resolution_clock::now();
+  std::map<std::string, std::unique_ptr<nic::InferContext::Result>> results;
+  status = wrap(context_->Run(&results), "evaluate(): unable to run and/or get result");
+  if (!status) {
+    this->output_.resize(nOutput_ * batchSize_, 0.f);
+    this->finish(false);
+    return;
+  }
+
+  auto t2 = std::chrono::high_resolution_clock::now();
+  if (!this->debugName_.empty())
+    edm::LogInfo(this->fullDebugName_) << "Remote time: "
+                                       << std::chrono::duration_cast<std::chrono::microseconds>(t2 - t1).count();
+
+  const auto& end_status = getServerSideStatus();
+
+  if (verbose()) {
+    const auto& stats = summarizeServerStats(start_status, end_status);
+    reportServerSideStats(stats);
+  }
+
+  status = getResults(*results.begin()->second);
+
+  this->finish(status);
+}
+
+//specialization for true async
+template <>
+void TritonClientAsync::evaluate() {
+  //in case there is nothing to process
+  if (batchSize_ == 0) {
+    this->finish(true);
+    return;
+  }
+
+  //common operations first
+  bool status = setup();
+  if (!status) {
+    this->finish(false);
+    return;
+  }
+
+  // Get the status of the server prior to the request being made.
+  const auto& start_status = getServerSideStatus();
+
+  //non-blocking call
+  auto t1 = std::chrono::high_resolution_clock::now();
+  status =
+      wrap(context_->AsyncRun([t1, start_status, this](nic::InferContext* ctx,

```

Comment:
out of curiosity: why do you prefer to measure `t1` outside the lambda and then capture it, instead of directly measuring it inside the lambda ?

the latter would also make it easier to measure it only if `(!debugName_.empty())`, since it is only used in that case.

Reply:
`AsyncRun()` performs some operations before executing the callback, and we would like to include those in our profiling.